src/Core/View.js

```diff
@@ -158,29 +144,19 @@ function _preprocessLayer(view, layer, provider, parentLayer) {
     }
 
     if (!layer.whenReady) {
-        let providerPreprocessing = Promise.resolve();
         if (provider && provider.preprocessDataLayer) {
-            providerPreprocessing = provider.preprocessDataLayer(layer, view, view.mainLoop.scheduler, parentLayer);
-            if (!(providerPreprocessing && providerPreprocessing.then)) {
-                providerPreprocessing = Promise.resolve();
-            }
-        } else if (layer.source) {
-            if (!layer.source.isSource) {
-                console.warn('Deprecation warning: passing a source as an object is deprecated. Instantiate the source before adding it to the layer instead.');
-                const protocol = layer.source.protocol;
-                layer.source = new (supportedSource.get(protocol))(layer.source, layer.projection);
-            }
-
-            providerPreprocessing = layer.source.whenReady || providerPreprocessing;
+            layer.whenReady = provider.preprocessDataLayer(layer, view, view.mainLoop.scheduler, parentLayer);
+        } else if (layer.source && layer.source.whenReady) {
```

Comment:
All sources should be have `whenReady`.
we could have:
```suggestion
        } else { 
            layer.whenReady = layer.source.whenReady;
        }
```

Reply:
Ok so we would have in `Source.js` constructor something like `whenReady = Promise.resolve();` ?

Reply:
yes

Reply:
In fact, we can't do that, because a layer doesn't always have a source.

Reply:
ok, you have a example?

Reply:
Yes, see the `3dtiles.html` example

Reply:
ok